components/tabs/src/sync/record.rs

```diff
@@ -27,7 +28,163 @@ pub struct TabsRecord {
+
+                TabsRecordTab {
+                    title,
+                    url_history,
+                    icon,
+                    last_used,
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let ttl: u32 = payload.data["ttl"].as_u64().unwrap_or_default() as u32;
+
+        Ok(TabsRecord {
+            id: payload.id.to_string(),
+            client_name,
+            tabs,
+            ttl,
+        })
+    }
+}
+
+fn parse_last_used(last_used_val: &JsonValue) -> u64 {
+    // In order to support older clients, we are handling `last used` values that
+    // are either floats, integers, or stringified floats or integers. We attempt to
+    // represent `last used` as a float before converting it to an integer. If that
+    // operation isn't successful, we try converting `last used` to an integer directly.
+    // If that isn't successful, the returned value will be zero.
+
+    if last_used_val.is_string() {
+        let l = last_used_val.as_str().unwrap_or_default();
+
+        match l.parse::<f64>() {
+            Ok(f) => f.trunc() as u64,
+            Err(_) => l.parse::<u64>().unwrap_or_default(),
+        }
+    } else {
+        match last_used_val.as_f64() {
+            Some(f) => f.trunc() as u64,
+            None => last_used_val.as_u64().unwrap_or_default(),
+        }
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    use super::*;
+    use serde_json::json;
+    use sync_guid::Guid;
+
+    #[test]
+    fn test_tabs_record_from_payload() -> Result<()> {
+        let guid = Guid::random();
+        let mut data = Map::new();
+
+        let tabs: JsonValue = json!([
+            {
+                "title": "Example",
+                "urlHistory": [
+                    "example.com",
+                    "example2.com"
+                ],
+                "icon": "example.png",
+                "lastUsed": 1623745123 // test with integer value
+            },
+            {
+                "title": "Test",
+                "urlHistory": [
+                    "test.com",
+                    "test2.com"
+                ],
+                "icon": "test.png",
+                "lastUsed": 1623745000.99 // test with float value
+            },
+            {
+                "title": "Example2",
+                "urlHistory": [
+                    "example2.com"
+                ],
+                "icon": "example2.png",
+                "lastUsed": "1623745144" // test with stringified integer value
+            },
+            {
+                "title": "Test2",
+                "urlHistory": [
+                    "test2.com"
+                ],
+                "icon": "test2.png",
+                "lastUsed": "1623745018.99" // test with stringified float value
+            },
+        ]);
+
+        data.insert("tabs".to_string(), tabs);
+        data.insert("clientName".to_string(), JsonValue::from("Nightly"));
+        data.insert("ttl".to_string(), JsonValue::from(0));
```

Comment:
I don't feel strongly about it, but I wonder if this would be more readable if you constructed the data wrapper as part of the JSON macro, along the lines of:

```
let data = json!({
  "clientName": ...,
  "tabs": [
     ...
   ],
}).as_map().unwrap().clone();
```